### ISSUE TITLE
problem with ExtractEdges.py

### ISSUE THREAD

**Issue by shinlin77** [asrhou/NATMI#47]
I get the following error.  The csv file is quite big.  When I break it up into 2 pieces, I don't get the error, but I think that might be a red herring (because other files just as big don't cause this error).  I don't think there is a problem with my pipeline per se, because other files I produce do work.  

```
/home/usr/new/src/anaconda3/bin/python ExtractEdges.py --species human --emFile /home/usr/new/encode/em_HF.csv --annFile /home/usr/new/encode/metadata_HF.csv --interDB lrc2a --out /home/usr/new/encode/NATMI_res/HF --coreNum 4
===================================================
Input data:
Species to obtain abundance data: human
The expression matrix file: /home/usr/new/encode/em_HF.csv
The metafile: /home/usr/new/encode/metadata_HF.csv
The ligand-receptor interaction database: lrc2a.txt
Species to obtain ligand-receptor interactions: human
The number of cores to use: 4
The folder to save all results: /home/usr/new/encode/NATMI_res/HF
===================================================



#### the folder "/home/usr/new/encode/NATMI_res/HF" has been created to save the analysis results
#### cluster the expression matrix
Traceback (most recent call last):
  File "ExtractEdges.py", line 558, in <module>
    main(species, opt.emFile, opt.annFile, idType, interDB, interSpecies, opt.coreNum, opt.out)
  File "ExtractEdges.py", line 464, in main
    sumEMDF, meanEMDF, countEMDF, cellEMDF = ClusterAnnotateEM(resultDir, em, ann)
  File "ExtractEdges.py", line 68, in ClusterAnnotateEM
    cellDF = clusterDF[clusterDF>0].count(axis=1).astype(float).to_frame(name=clusterId)
  File "/home/usr/new/src/anaconda3/lib/python3.7/site-packages/pandas/core/ops/common.py", line 65, in new_method
    return method(self, other)
  File "/home/usr/new/src/anaconda3/lib/python3.7/site-packages/pandas/core/arraylike.py", line 45, in __gt__
    return self._cmp_method(other, operator.gt)
  File "/home/usr/new/src/anaconda3/lib/python3.7/site-packages/pandas/core/frame.py", line 5968, in _cmp_method
    new_data = self._dispatch_frame_op(other, op, axis=axis)
  File "/home/usr/new/src/anaconda3/lib/python3.7/site-packages/pandas/core/frame.py", line 6005, in _dispatch_frame_op
    bm = self._mgr.apply(array_op, right=right)
  File "/home/usr/new/src/anaconda3/lib/python3.7/site-packages/pandas/core/internals/managers.py", line 425, in apply
    applied = b.apply(f, **kwargs)
  File "/home/usr/new/src/anaconda3/lib/python3.7/site-packages/pandas/core/internals/blocks.py", line 378, in apply
    result = func(self.values, **kwargs)
  File "/home/usr/new/src/anaconda3/lib/python3.7/site-packages/pandas/core/ops/array_ops.py", line 243, in comparison_op
    res_values = comp_method_OBJECT_ARRAY(op, lvalues, rvalues)
  File "/home/usr/new/src/anaconda3/lib/python3.7/site-packages/pandas/core/ops/array_ops.py", line 55, in comp_method_OBJECT_ARRAY
    result = libops.scalar_compare(x.ravel(), y, op)
  File "pandas/_libs/ops.pyx", line 103, in pandas._libs.ops.scalar_compare
TypeError: '>' not supported between instances of 'str' and 'int'
```

**Comment by tyronechen** [comment]
Hi, just adding on that i also replicated the same error:

```
ExtractEdges.py --species zebrafish \
  --emFile "/██████/data/foo.csv" \
  --annFile "/██████/data/foo.metadata.csv" \
  --out "/██████/results/foo" \
  --coreNum 16

===================================================
Input data:
Species to obtain abundance data: zebrafish
The expression matrix file: /██████/natmi//data/foo.csv
The metafile: /██████/natmi//data/foo.metadata.csv
The ligand-receptor interaction database: lrc2p.csv
Species to obtain ligand-receptor interactions: human
The number of cores to use: 16
The folder to save all results: /██████/natmi/results/foo
===================================================
#### the folder "/██████/natmi/results/foo" has been created to save the analysis results
#### cluster the expression matrix
Traceback (most recent call last):
  File "/██████/NATMI/ExtractEdges.py", line 558, in <module>
    main(species, opt.emFile, opt.annFile, idType, interDB, interSpecies, opt.coreNum, opt.out)
  File "/██████/NATMI/ExtractEdges.py", line 464, in main
    sumEMDF, meanEMDF, countEMDF, cellEMDF = ClusterAnnotateEM(resultDir, em, ann)
  File "/██████/NATMI/ExtractEdges.py", line 68, in ClusterAnnotateEM
    cellDF = clusterDF[clusterDF>0].count(axis=1).astype(float).to_frame(name=clusterId)
  File "/██████/NATMI/natmi_lib/lib/python3.7/site-packages/pandas/core/ops/__init__.py", line 852, in f
    new_data = dispatch_to_series(self, other, op)
  File "/██████/NATMI/natmi_lib/lib/python3.7/site-packages/pandas/core/ops/__init__.py", line 381, in dispatch_to_series
    bm = left._data.apply(array_op, right=right)
  File "/██████/NATMI/natmi_lib/lib/python3.7/site-packages/pandas/core/internals/managers.py", line 440, in apply
    applied = b.apply(f, **kwargs)
  File "/██████/NATMI/natmi_lib/lib/python3.7/site-packages
/pandas/core/ops/__init__.py", line 381, in dispatch_to_series
    bm = left._data.apply(array_op, right=right)
  File "/██████/NATMI/natmi_lib/lib/python3.7/site-packages
/pandas/core/internals/managers.py", line 440, in apply
    applied = b.apply(f, **kwargs)
  File "/██████/NATMI/natmi_lib/lib/python3.7/site-packages/pandas/core/internals/blocks.py", line 390, in apply
    result = func(self.values, **kwargs)
  File "/██████/NATMI/natmi_lib/lib/python3.7/site-packages/pandas/core/ops/array_ops.py", line 247, in comparison_op
    res_values = comp_method_OBJECT_ARRAY(op, lvalues, rvalues)
  File "/██████/NATMI/natmi_lib/lib/python3.7/site-packages/pandas/core/ops/array_ops.py", line 57, in comp_method_OBJECT_ARRAY
    result = libops.scalar_compare(x.ravel(), y, op)
  File "pandas/_libs/ops.pyx", line 96, in pandas._libs.ops.scalar_compare
TypeError: '>' not supported between instances of 'str' and 'int'
```

**Comment by mcjmigdal** [comment]
Hi, we run into the exactly same error.

```
===================================================
Input data:
Species to obtain abundance data: zebrafish
The expression matrix file: /root/data/int_all_exp_matrix.fixed.csv
The metafile: /root/data/int_all_metadata.csv
The ligand-receptor interaction database: lrc2a.txt
Species to obtain ligand-receptor interactions: human
The number of cores to use: 14
===================================================
#### the folder "/root/data/int_all_exp_matrix.fixed" has been created to save the analysis results
#### cluster the expression matrix   
Traceback (most recent call last):   
  File "ExtractEdges.py", line 560, in <module>
    main(species, opt.emFile, opt.annFile, idType, interDB, interSpecies, opt.coreNum, opt.out)
  File "ExtractEdges.py", line 465, in main
    sumEMDF, meanEMDF, countEMDF, cellEMDF = ClusterAnnotateEM(resultDir, em, ann)
  File "ExtractEdges.py", line 69, in ClusterAnnotateEM
    cellDF = clusterDF[clusterDF>0].count(axis=1).astype(float).to_frame(name=clusterId)
  File "/opt/conda/lib/python3.7/site-packages/pandas/core/ops/__init__.py", line 852, in f
    new_data = dispatch_to_series(self, other, op)
  File "/opt/conda/lib/python3.7/site-packages/pandas/core/ops/__init__.py", line 381, in dispatch_to_series
    bm = left._data.apply(array_op, right=right)
  File "/opt/conda/lib/python3.7/site-packages/pandas/core/internals/managers.py", line 440, in apply
    applied = b.apply(f, **kwargs)   
  File "/opt/conda/lib/python3.7/site-packages/pandas/core/internals/blocks.py", line 390, in apply
    result = func(self.values, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/pandas/core/ops/array_ops.py", line 247, in comparison_op
    res_values = comp_method_OBJECT_ARRAY(op, lvalues, rvalues)
  File "/opt/conda/lib/python3.7/site-packages/pandas/core/ops/array_ops.py", line 57, in comp_method_OBJECT_ARRAY
    result = libops.scalar_compare(x.ravel(), y, op)
  File "pandas/_libs/ops.pyx", line 96, in pandas._libs.ops.scalar_compare
TypeError: '>' not supported between instances of 'str' and 'int'
```

as we are trying to understand what is going on, I found that our expression matrix is read by pandas with mixed data types especially some of the number are being load as strings. Further we found that some of the numbers in our expression matrix are extremely high (eg. 63984349353005490176) which I suspect might be causing pandas converting them to stings.

In our analysis we are preparing input data from Seurat using command for Seurat 3.X

```
write.csv(100 * (exp(as.matrix(GetAssayData(object = object, assay = "RNA", slot = "data"))) - 1), "em.csv", row.names = T) # Seurat 3.X
```

However, we are in fact using Seurat 4.X which we suspect to be the reason of getting such a high numbers in the output data matrix. In our Seurat object the matrix under assay=RNA slot=data is exactly the same as the one stored under assay=RNA slot=counts. However, the README says  assay=RNA slot=data contains log normalized TPM/CPM.

In conclusion I think that to use Seurat 4.X we need a new command for generating TPM/CPM input expression matrix.

**Comment by asrhou** [comment]
Hi guys,

Sorry for the late reply. If the expression matrix is too big, is it possible to use tools such as Metacell to reduce it?

Many thanks,
Rui
